crypto/evp/e_aes.c

```diff
@@ -1076,54 +1125,220 @@ static int s390x_aes_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key,
 # define S390X_aes_128_cbc_CAPABLE	1	/* checked by callee */
 # define S390X_aes_192_cbc_CAPABLE	1
 # define S390X_aes_256_cbc_CAPABLE	1
+# define S390X_AES_CBC_CTX		EVP_AES_KEY
+
+# define s390x_aes_cbc_init_key aes_init_key
 
 # define s390x_aes_cbc_cipher aes_cbc_cipher
 static int s390x_aes_cbc_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
                                 const unsigned char *in, size_t len);
 
-# define S390X_aes_128_ecb_CAPABLE	0
-# define S390X_aes_192_ecb_CAPABLE	0
-# define S390X_aes_256_ecb_CAPABLE	0
+# define S390X_aes_128_ecb_CAPABLE	S390X_aes_128_CAPABLE
+# define S390X_aes_192_ecb_CAPABLE	S390X_aes_192_CAPABLE
+# define S390X_aes_256_ecb_CAPABLE	S390X_aes_256_CAPABLE
+
+static int s390x_aes_ecb_init_key(EVP_CIPHER_CTX *ctx,
+                                  const unsigned char *key,
+                                  const unsigned char *iv, int enc)
+{
+    S390X_AES_ECB_CTX *cctx = EVP_C_DATA(S390X_AES_ECB_CTX, ctx);
+    const int keylen = EVP_CIPHER_CTX_key_length(ctx);
+
+    memcpy(cctx->km.param.k, key, keylen);
+    cctx->fc = S390X_AES_FC(keylen);
+    return 1;
+}
 
-# define s390x_aes_ecb_cipher aes_ecb_cipher
 static int s390x_aes_ecb_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
-                                const unsigned char *in, size_t len);
+                                const unsigned char *in, size_t len)
+{
+    S390X_AES_ECB_CTX *cctx = EVP_C_DATA(S390X_AES_ECB_CTX, ctx);
+    const int enc = EVP_CIPHER_CTX_encrypting(ctx);
+    const int fc_flags = enc ? 0 : S390X_DECRYPT;
 
-# define S390X_aes_128_ofb_CAPABLE	0
-# define S390X_aes_192_ofb_CAPABLE	0
-# define S390X_aes_256_ofb_CAPABLE	0
+    s390x_km(in, len, out, cctx->fc | fc_flags, &cctx->km.param);
+    return 1;
+}
+
+# define S390X_aes_128_ofb_CAPABLE (S390X_aes_128_CAPABLE &&		\
```

Comment:
Well, if you omitted \*_cfb8_CAPABLE, then why keep these [and \*_cfb_CAPABLE]?

Reply:
I omitted S390X_aes_xxx_CAPABLE in case of cfb8 only, because cfb8 does not need km function code to process partial block.

Reply:
My impression was that you can't have km extensions without km. The chain of checks OPENSSL_s390x_facilities suggests it, i.e. if there is no message-security-assist facility, it won't check for message-security-assist-4 one. But ok, fair enough...

Reply:
My understanding of the feature detection architecture is the following:

The facility bits stored by stfle instruction indicate the availability of msa and msa-extensions.

Dependence between msa-extensions is the following (from current POP 1-18 to 1-22): msa-ex1, msa-ex2, msa-ex3 (bit 76), and msa-ex6 require the msa (bit 17). msa-ex4 (bit 77) requires msa-ex3. msa-ex5 (bit 57) requires msa-ex2's sha-512 capabilities. msa-ex7 requires msa-ex5. POP does not list any dependencies for msa-ex8 (bit 146) in the cited chapters. However, on 4-101 it says "Bit 76 is one when bit 146 is one.", i.e., msa-ex8 requires msa-ex3. ("a requires b" in this context means, that availability of a implies availability of b).

msa levels which introduce new intructions have a corresponding facility bit, those which only add new function codes to instructions dont. A msa level's facility bit guarantees the availability of at least function code 0 (query) for the intruction(s) introduced on that level.

Bottom line is that msa extensions dont seem to be "strictly inclusive" in the sense that availability of a msa-extension level implies to availability of all smaller msa-extension levels. However, i havent seen a real machine that implements non-inclusive msa levels so i think its okay to stop msa level detection once the first not installed level is detected. Another thing is that i would not make any assumptions about an instruction's installed function codes without checking query/function code 0 return value. In general i consider its always save to (1) check stfle return value for msa indications, (2) check the instructions' query (which are available with the detected msa levels) for function codes.

Reply:
> availability of a msa-extension level implies to availability of all smaller msa-extension levels.

I'm just looking at OPENSSL_s390x_facilities, and it's arranged as if there is such relationship. If there is no base message-security-assist detected, it does not look for *any* message-security-assist. And if there is no 4 it doesn't look for 5, and if there is no 5 it doesn't look for 8... Well, something to ponder over later on...

Reply:
Yes, im aware of the OPENSSL_s390x_facilities code, but as i said, though the architecture seems to allow "non-inclusive" msa levels (i cant find any statement supporting inclusive levels and listing dependencies between levels explicitly is an indication against it), i dont think there are actual implementations (apart from virtualization) of the architecture which use that freedom.